test_conformance/api/test_kernel_attributes.cpp

```diff
@@ -0,0 +1,322 @@
+        &vect_tests,           &work_tests,           &reqd_tests,
+
+        &work_vect_tests,      &vect_reqd_tests,      &reqd_work_tests,
+
+        &vect_work_reqd_tests, &work_reqd_vect_tests, &reqd_vect_work_tests
+    };
+    return all_tests;
+}
+
+static const std::vector<AttributePermutations*>
+initialise_attribute_data(cl_device_id deviceID)
+{
+    /* This vector stores different work group dimensions that can be used by
+     * the reqd_work_group_size and work_group_size_hint attributes. It
+     * currently only has a single value to minimise time complexity of the
+     * overall test but can be easily changed. */
+    static const std::vector<WorkGroupDimensions> work_group_dimensions = {
+        { 1, 1, 1 }
+    };
+    KernelAttributes vec_type_hint_data = generate_vec_type_hint_data(deviceID);
+    KernelAttributes work_group_size_hint_data =
+        generate_work_group_size_data(work_group_dimensions);
+    KernelAttributes reqd_work_group_size_data =
+        generate_reqd_work_group_size_data(work_group_dimensions);
+
+    /* Generate all the permutations of attributes to create different test
+     * suites */
+    return generate_attribute_tests(vec_type_hint_data,
+                                    work_group_size_hint_data,
+                                    reqd_work_group_size_data);
+}
+
+static bool run_test(cl_context context, cl_device_id deviceID,
+                     const AttributePermutations& permutations)
+{
+    bool success = true;
+    for (auto attribute_permutation : permutations)
+    {
+
+        std::string kernel_source_string =
+            generate_kernel_source(attribute_permutation);
+        const char* kernel_src = kernel_source_string.c_str();
+        clProgramWrapper program;
+        clKernelWrapper kernel;
+        cl_int err = create_single_kernel_helper(context, &program, &kernel, 1,
+                                                 &kernel_src, "test_kernel");
+        ASSERT_SUCCESS(err, "create_single_kernel_helper");
+
+        /* Get the size of the kernel attribute string returned */
+        size_t size = 0;
+        err = clGetKernelInfo(kernel, CL_KERNEL_ATTRIBUTES, 0, nullptr, &size);
+        ASSERT_SUCCESS(err, "clGetKernelInfo");
+        std::vector<char> attributes(size);
+        err = clGetKernelInfo(kernel, CL_KERNEL_ATTRIBUTES, attributes.size(),
+                              attributes.data(), nullptr);
+        ASSERT_SUCCESS(err, "clGetKernelInfo");
+        std::string attribute_string(attributes.data());
+        attribute_string.erase(
+            std::remove(attribute_string.begin(), attribute_string.end(), ' '),
+            attribute_string.end());
+        for (auto attribute : attribute_permutation)
+        {
+            if (attribute_string.find(attribute) == std::string::npos)
+            {
+                success = false;
+                log_error("ERR: expected: %s\nFound: %s\n", attribute.c_str(),
```

Comment:
This output could be made a little clearer.  Would suggest `"ERROR: did not find expected attribute: '%s'\n"`, with the full attribute list printed just once if a failure is found for the permutation, with a message such as `"Attribute list reported is: '%s'\n"`.

I would note that this doesn't raise an error if extra attributes were reported that were not in the permutation. Should it?

Reply:
The most recent changes look good to me. I see now that the `CL_KERNEL_ATTRIBUTES` info is not trivial to parse. I think it would be good to check that there are no extraneous attributes, but this could be added some other time.

Would be good to have independent review of the rest of this PR.